### PR TITLE
[RHCLOUD-23141] subscribe emails at event type level

### PR DESCRIPTION
**Pull request by g-duval**

### CODE REVIEW

File: aggregator/src/main/java/com/redhat/cloud/notifications/db/EmailAggregationRepository.java [hunk 1]
```diff
@@ -34,23 +39,27 @@ public List<AggregationCommand> getApplicationsWithPendingAggregationAccordinfOr
         List<Object[]> records = hqlQuery.getResultList();
         return records.stream()
                 .map(emailAggregationRecord -> new AggregationCommand(
-                    new EmailAggregationKey((String) emailAggregationRecord[0], (String) emailAggregationRecord[1], (String) emailAggregationRecord[2]),
-                    (LocalDateTime) emailAggregationRecord[3],
+                    new EmailAggregationKey((String) emailAggregationRecord[0], (String) emailAggregationRecord[1], (String) emailAggregationRecord[2], (String) emailAggregationRecord[2]),
```
Comment (josejulio) [inline aggregator/src/main/java/com/redhat/cloud/notifications/db/EmailAggregationRepository.java:42]: I guess it's a copy/paste typo :D 

```suggestion
                    new EmailAggregationKey((String) emailAggregationRecord[0], (String) emailAggregationRecord[1], (String) emailAggregationRecord[2], (String) emailAggregationRecord[3]),
```

Comment (g-duval) [reply aggregator/src/main/java/com/redhat/cloud/notifications/db/EmailAggregationRepository.java:42]: Yes it is, I fixed it on local, I will wait to go further on my end to end validation tests before pushing an update :)

Comment (josejulio) [reply aggregator/src/main/java/com/redhat/cloud/notifications/db/EmailAggregationRepository.java:42]: I just saw it and I know how frustrating can be these kind of typos so i had to tell :-)

File: aggregator/src/main/resources/application.properties [hunk 2]
```diff
@@ -1,3 +1,6 @@
+# Kafka bootstrap applies to all topics
+kafka.bootstrap.servers=localhost:9092
```
Comment (g-duval) [inline aggregator/src/main/resources/application.properties:2]: Fix kafka instance on local env to use same insatance as engine

**Comment by g-duval** [comment]
Migration script will be done on separate PR when we will be ready to deploy it on prod

**Review by gwenneg — COMMENTED** [review]
As we discussed on Meet, the changes around the aggregation process need to be fine-turned. Here are a few minor comments that I had already written before our discussion.

File: backend/src/main/java/com/redhat/cloud/notifications/routers/models/SettingsValueByEventTypeJsonForm.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+@JsonAutoDetect(fieldVisibility = Visibility.ANY)
+public class SettingsValueByEventTypeJsonForm {
+
+    private static final ObjectMapper mapper = new ObjectMapper();
```
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/routers/models/SettingsValueByEventTypeJsonForm.java:16]: Is this needed?

File: backend/src/main/java/com/redhat/cloud/notifications/routers/models/SettingsValueByEventTypeJsonForm.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                field.component = COMPONENT_SUBSCRIPTION;
+                switch (emailSubscriptionType) {
+                    case DAILY:
+                        field.label = "Daily digest";
+                        field.description =  "Daily summary of triggered application events in 24 hours span. See notification settings for configuration.";
+                        break;
+                    case INSTANT:
+                        field.label = "Instant notification";
+                        field.description = "Immediate email for each triggered application event. See notification settings for configuration.";
+                        field.checkedWarning = "Opting into this notification may result in a large number of emails";
```
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/routers/models/SettingsValueByEventTypeJsonForm.java:98]: The following line is present in the old class but not in this one. Is that intended?
```java
field.infoMessage = "You may still receive forced notifications for this service";
```

File: backend/src/main/java/com/redhat/cloud/notifications/routers/models/SettingsValueByEventTypeJsonForm.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    public Map<String, Application> bundles = new HashMap<>();
+
+    public static EventTypes fromSettingsValueEventTypes(SettingsValuesByEventType eventTypeEmailSubscriptions, String bundleName, String applicationName) {
+        SettingsValuesByEventType.ApplicationSettingsValue applicationSettingsValue = eventTypeEmailSubscriptions.bundles.get(bundleName).applications.get(applicationName);
+        return buildEventTypes(bundleName, applicationName, applicationSettingsValue);
+    }
+
+    public static SettingsValueByEventTypeJsonForm fromSettingsValue(SettingsValuesByEventType values) {
```
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/routers/models/SettingsValueByEventTypeJsonForm.java:62]: The old class contained some logic based on the `Sections` class. It seems to be gone here. Is that intended?

File: backend/src/main/java/com/redhat/cloud/notifications/routers/UserConfigResource.java [hunk 6]
```diff
@@ -201,10 +218,186 @@ private String settingsValuesToJsonForm(SettingsValues settingsValues) {
...
+    }
+
+    @GET
+    @Path("/notification-event-type-preference")
+    @Produces(APPLICATION_JSON)
+    @Operation(hidden = true)
+    public Response getSettingsSchemaByEventType(@Context SecurityContext sec) {
+
+        final RhIdPrincipal principal = (RhIdPrincipal) sec.getUserPrincipal();
+        final String name = principal.getName();
```
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/routers/UserConfigResource.java:276]:
```suggestion
        final String name = getUserName(sec);
```

File: common-aggregator/src/main/java/com/redhat/cloud/notifications/models/EmailAggregation.java [hunk 7]
```diff
@@ -42,6 +42,10 @@ public class EmailAggregation extends CreationTimestamped {
     @JsonProperty("application")
     private String applicationName;
 
+    @Size(max = 255)
+    @JsonProperty("event_type")
+    private String eventType;
```
[the remark this answers is not in the document]
Comment (g-duval) [reply common-aggregator/src/main/java/com/redhat/cloud/notifications/models/EmailAggregation.java]: No you are right

File: database/src/main/resources/db/migration/V1.71.0__RHCLOUD-21306_ability_to_subscribe_unsubscribe_down_to_the_event_type.sql [hunk 8]
```diff
@@ -0,0 +1,17 @@
+CREATE TABLE public.email_subscriptions (
+    account_id varchar(50) NULL,
```
[the remark this answers is not in the document]
Comment (g-duval) [reply database/src/main/resources/db/migration/V1.71.0__RHCLOUD-21306_ability_to_subscribe_unsubscribe_down_to_the_event_type.sql]: Ok

File: database/src/main/resources/db/migration/V1.72.0__RHCLOUD-21306_ability_to_subscribe_unsubscribe_down_to_the_event_type.sql [hunk 9]
```diff
@@ -0,0 +1,11 @@
+CREATE TABLE public.email_subscriptions (
```
Comment (gwenneg) [inline database/src/main/resources/db/migration/V1.72.0__RHCLOUD-21306_ability_to_subscribe_unsubscribe_down_to_the_event_type.sql:1]: How and when will the data be migrated from the old table to this one?

File: backend/src/main/java/com/redhat/cloud/notifications/routers/models/SettingsValueByEventTypeJsonForm.java [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        @JsonInclude(Include.NON_NULL)
+        public Object initialValue;
+        public String component;
+        @JsonInclude(Include.NON_NULL)
+        public List<Object> validate;
+        @JsonInclude(Include.NON_NULL)
+        public List<Field> fields;
+        @JsonInclude(Include.NON_NULL)
+        public String checkedWarning;
+        public String infoMessage;
```
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/routers/models/SettingsValueByEventTypeJsonForm.java:33]:
```suggestion
        @JsonInclude(Include.NON_NULL)
        public String infoMessage;
```

File: database/src/main/resources/db/migration/V1.72.0__RHCLOUD-21306_ability_to_subscribe_unsubscribe_down_to_the_event_type.sql (hunk 9 above)
Comment (g-duval) [reply database/src/main/resources/db/migration/V1.72.0__RHCLOUD-21306_ability_to_subscribe_unsubscribe_down_to_the_event_type.sql:1]: By simple sql request when we (back + front team) will be ready to use this new mechanismon prod

Comment (gwenneg) [reply database/src/main/resources/db/migration/V1.72.0__RHCLOUD-21306_ability_to_subscribe_unsubscribe_down_to_the_event_type.sql:1]: Features enabled with feature flags are usually migrated using a REST endpoint to make sure everything is enabled/migrated at the right time. Let's talk about that later and open a separate PR about it.

File: backend/src/main/java/com/redhat/cloud/notifications/routers/models/SettingsValuesByEventType.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    public static class ApplicationSettingsValue {
+        @JsonIgnore
+        public String displayName;
+        public Map<String, EventTypeSettingsValue> eventTypes = new HashMap<>();
+    }
+
+    public static class EventTypeSettingsValue {
+        @JsonIgnore
+        public String displayName;
+        public Map<EmailSubscriptionType, Boolean> emailSubscriptionTypes = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (josejulio) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/models/SettingsValuesByEventType.java:34]: I think the new name is good - it's more specific.

File: backend/src/main/java/com/redhat/cloud/notifications/routers/UserConfigResource.java [hunk 12]
```diff
@@ -201,10 +220,191 @@ private String settingsValuesToJsonForm(SettingsValues settingsValues) {
...
+
+    private SettingsValuesByEventType getSettingsValueForUserByEventType(List<EventTypeEmailSubscription> emailSubscriptions, String orgId) {
+        SettingsValuesByEventType settingsValues = new SettingsValuesByEventType();
+
+        for (Bundle bundle : bundleRepository.getBundles()) {
+            List<Application> applicationsWithForcedEmails = applicationRepository.getApplicationsWithForcedEmail(bundle.getId(), orgId);
+            List<String> listApplicationsWithForcedEmails = applicationsWithForcedEmails.stream().map(app -> app.getName()).collect(Collectors.toList());
+            for (Application application : bundle.getApplications()) {
+                addApplicationStructureDetails(settingsValues, application, listApplicationsWithForcedEmails.contains(application.getName()));
+            }
```
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/routers/UserConfigResource.java:404]:
```suggestion
            List<String> applicationsWithForcedEmails = applicationRepository.getApplicationsWithForcedEmail(bundle.getId(), orgId)
                    .stream().map(Application::getName).collect(Collectors.toList());
            for (Application application : bundle.getApplications()) {
                addApplicationStructureDetails(settingsValues, application, applicationsWithForcedEmails.contains(application.getName()));
            }
```

File: backend/src/test/java/com/redhat/cloud/notifications/routers/UserConfigResourceTest.java [hunk 13]
```diff
@@ -56,6 +68,9 @@ void beforeEach() {
     @RestClient
     TemplateEngineClient templateEngineClient;
 
+    @InjectSpy
```
Comment (gwenneg) [inline backend/src/test/java/com/redhat/cloud/notifications/routers/UserConfigResourceTest.java:71]: Do you need a Mockito spy for a specific reason? If not, then we usually use `@InjectMock`.

File: backend/src/test/java/com/redhat/cloud/notifications/routers/UserConfigResourceTest.java [hunk 14]
```diff
@@ -255,7 +312,7 @@ void testSettings() {
         assertEquals(true, preferences.getDailyEmail());
         assertEquals(true, preferences.getInstantEmail());
 
-        // does not fail if we have unknown apps in our bundle's settings
+        // Fail if we have unknown apps in our bundle's settings
```
Comment (gwenneg) [inline backend/src/test/java/com/redhat/cloud/notifications/routers/UserConfigResourceTest.java:315]: Does it really fail? The HTTP status in the response below seems to be `200`.

File: backend/src/test/java/com/redhat/cloud/notifications/routers/UserConfigResourceTest.java [hunk 15]
```diff
@@ -320,4 +377,161 @@ void testSettings() {
...
+        // Daily to true
+        updateAndCheckUserPreference(path, identityHeader, bundle, application, eventType, true, false);
+
+        // Instant to true
+        updateAndCheckUserPreference(path, identityHeader, bundle, application, eventType, false, true);
+
+        // Both to true
+        updateAndCheckUserPreference(path, identityHeader, bundle, application, eventType, true, true);
+
+        // Fail if we have unknown apps in our bundle's settings
```
Comment (gwenneg) [inline backend/src/test/java/com/redhat/cloud/notifications/routers/UserConfigResourceTest.java:436]: Is it an unknown app or unknown event type?

File: backend/src/test/java/com/redhat/cloud/notifications/routers/UserConfigResourceTest.java [hunk 16]
```diff
@@ -320,4 +377,161 @@ void testSettings() {
...
+        assertThrows(PersistenceException.class, () -> {
+            emailSubscriptionRepository.subscribeEventType(orgId, username, UUID.randomUUID(), DAILY);
+        });
+
+        // not fail if we have unknown apps in our bundle's settings on unsubscibe, but nb affected rows must be 0
```
Comment (gwenneg) [inline backend/src/test/java/com/redhat/cloud/notifications/routers/UserConfigResourceTest.java:441]: Same question.

**Review by gwenneg — COMMENTED** [review]
`UserConfigResource` and its tests were quite hard to review. You'll find suggestions that make some parts of the code a bit simpler. Feel free to commit or ignore them, they are not mandatory!

File: database/src/main/resources/db/migration/V1.72.0__RHCLOUD-21306_ability_to_subscribe_unsubscribe_down_to_the_event_type.sql (hunk 9 above)
Comment (g-duval) [reply database/src/main/resources/db/migration/V1.72.0__RHCLOUD-21306_ability_to_subscribe_unsubscribe_down_to_the_event_type.sql:1]: Ok I created ticket RHCLOUD-24352 to track this point

File: engine/src/main/java/com/redhat/cloud/notifications/processors/email/EmailAggregator.java [hunk 17]
```diff
@@ -42,18 +43,45 @@ public class EmailAggregator {
     @Inject
     EmailSubscriptionRepository emailSubscriptionRepository;
 
+    @Inject
+    FeatureFlipper featureFlipper;
+
```
[the remark this answers is not in the document]
Comment (g-duval) [reply engine/src/main/java/com/redhat/cloud/notifications/processors/email/EmailAggregator.java:48]: Ok I added a test on it

File: backend/src/main/java/com/redhat/cloud/notifications/routers/UserConfigResource.java [hunk 18]
```diff
@@ -201,10 +220,191 @@ private String settingsValuesToJsonForm(SettingsValues settingsValues) {
...
+    }
+
+
+    @POST
+    @Path("/notification-event-type-preference")
+    @Consumes(APPLICATION_JSON)
+    @Produces(TEXT_PLAIN)
+    @Operation(hidden = true)
+    @Transactional
+    public Response saveSettingsByEventType(@Context SecurityContext sec, @NotNull @Valid SettingsValuesByEventType userSettings) {
```
[the remark this answers is not in the document]
Comment (gwenneg) [reply backend/src/main/java/com/redhat/cloud/notifications/routers/UserConfigResource.java:240]: I forgot a "could" in my comment but you understood what I meant :smile:

File: engine/src/test/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessorTest.java [hunk 19]
```diff
@@ -93,6 +93,7 @@ class EmailSubscriptionTypeProcessorTest {
     void beforeEach() {
         featureFlipper.setUseTemplatesFromDb(false);
         featureFlipper.setUseDefaultTemplate(false);
+        featureFlipper.setUseEventTypeForSubscriptionEnabled(true);
```
[the remark this answers is not in the document]
Comment (gwenneg) [reply engine/src/test/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessorTest.java:96]:
```suggestion
```

**Comment by gwenneg** [comment]
FYI I'm not merging on purpose. @g-duval will do it later.

**Pull request merged**